geth/swarm.go

```diff
@@ -0,0 +1,186 @@
+package geth
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/contracts/chequebook"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm"
+	bzzapi "github.com/ethereum/go-ethereum/swarm/api"
+	"path/filepath"
+	"sync"
+)
+
+var (
+	ErrInvalidSwarmService           = errors.New("swarm service is unavailable")
+	ErrSwarmIdentityInjectionFailure = errors.New("failed to inject identity into Swarm")
+)
+
+type SwarmService struct {
+	swarm  *swarm.Swarm
+	keys   map[string]*ecdsa.PrivateKey
+	keysMu sync.RWMutex // Mutex to sync identity keys
+}
+
+func NewSwarmService() *SwarmService {
+	service := &SwarmService{
+		keys: make(map[string]*ecdsa.PrivateKey),
+	}
+
+	return service
+}
+
+// activateEthService configures and registers the eth.Ethereum service with a given node.
+func (self *SwarmService) StartSwarm(prvkey *ecdsa.PrivateKey) error {
+	if self.swarm != nil {
+		err := self.Stop()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	nm := NodeManagerInstance()
+	stack := nm.node.geth
+
+	chbookaddr := common.Address{}
+	bzzdir := filepath.Join(stack.InstanceDir(), "swarm")
+
+	networkId := uint64(1)
+	if UseTestnet {
+		networkId = 3
+	}
+
+	bzzconfig, err := bzzapi.NewConfig(bzzdir, chbookaddr, prvkey, networkId)
+	if err != nil {
+		return fmt.Errorf("%v: %v", err, err)
+	}
+
+	var client *ethclient.Client
+	client, err = ethclient.Dial(fmt.Sprintf("http://%s", stack.HTTPEndpoint()))
+	if err != nil {
+		return fmt.Errorf("Can't connect: %v", err)
+	}
+
+	self.swarm, err = swarm.NewSwarm(nil, client, bzzconfig, false, true, "")
+	if err != nil {
+		return err
+	}
+
+	return self.Start(nm.node.geth.Server())
+}
+
+func (self *SwarmService) Start(net *p2p.Server) error {
+	if self.swarm != nil && net != nil {
+		return self.swarm.Start(net)
+	}
+
+	return nil
+}
+
+func (self *SwarmService) Stop() error {
+	if self.swarm != nil {
+		err := self.swarm.Stop()
+		if err != nil {
+			return err
+		}
+
+		self.swarm = nil
+	}
+
+	return nil
+}
+
+func (self *SwarmService) Protocols() []p2p.Protocol {
```

Comment:
- well, if you need to manage `swarm.Swarm` object yourself (which you do in `StartSwarm()` via calling `NewSwarm()` and producing new instance of the service everytime), then probably you do not need to `activateSwarmService()` at all?
- At least there's little to no point passing `stack.Register()` anything that you need to replace on every account select (be it in form of proxy or direct `swarm.Swarm` instance)
- so, since you are already managing `swarm.Swarm` creation and swapping, then probably just not add it to stack at all?